creusot/tests/should_succeed/result/own.rs

```diff
@@ -0,0 +1,213 @@
+extern crate creusot_contracts;
+
+use creusot_contracts::*;
+
+pub enum OwnResult<T, E> {
+    Ok(T),
+    Err(E),
+}
+
+impl<T, E> OwnResult<T, E> {
+    #[ensures(result == exists<t: T> *self == OwnResult::Ok(t))]
+    pub fn is_ok(&self) -> bool {
+        matches!(*self, OwnResult::Ok(_))
+    }
+
+    #[ensures(result == exists<e: E> *self == OwnResult::Err(e))]
+    pub fn is_err(&self) -> bool {
+        !self.is_ok()
+    }
+
+    #[ensures(forall<t: T> self == OwnResult::Ok(t) ==> result == Some(t))]
+    #[ensures((exists<e: E> self == OwnResult::Err(e)) ==> result == None)]
+    pub fn ok(self) -> Option<T> {
+        match self {
+            OwnResult::Ok(x) => Some(x),
+            #[allow(unused_variables)]
+            OwnResult::Err(x) => None,
+        }
+    }
+
+    #[ensures((exists<t: T> self == OwnResult::Ok(t)) ==> result == None)]
+    #[ensures(forall<e: E> self == OwnResult::Err(e) ==> result == Some(e))]
+    pub fn err(self) -> Option<E> {
+        match self {
+            #[allow(unused_variables)]
+            OwnResult::Ok(x) => None,
+            OwnResult::Err(x) => Some(x),
+        }
+    }
+
+    #[ensures(forall<t: &T> *self == OwnResult::Ok(*t) ==> result == OwnResult::Ok(t))]
+    #[ensures(forall<e: &E> *self == OwnResult::Err(*e) ==> result == OwnResult::Err(e))]
+    #[ensures((exists<t: &T> *self == OwnResult::Ok(*t))|| (exists<e: &E> *self == OwnResult::Err(*e)))]
+    pub fn as_ref(&self) -> OwnResult<&T, &E> {
+        match *self {
+            OwnResult::Ok(ref x) => OwnResult::Ok(x),
+            OwnResult::Err(ref x) => OwnResult::Err(x),
+        }
+    }
+
+    #[ensures(
+        forall<t: &mut T> *self == OwnResult::Ok(*t) ==> ^self == OwnResult::Ok(^t) && result == OwnResult::Ok(t)
+    )]
+    #[ensures(
+        forall<e: &mut E> *self == OwnResult::Err(*e) ==> ^self == OwnResult::Err(^e) && result == OwnResult::Err(e)
+    )]
```

Comment:
Change the `&&` to a `==>`

Reply:
That does work here, but when doing it to the standard `Result` spec it doesn't verify again.

Reply:
Well the `&&` is just wrong, so it needs to be done, I'm not sure what you mean by it not working for the standard result spec, do you mean that you can't verify client code afterwards?

Reply:
Yes, the "result/result" test does not fully verify.

Reply:
Actually, I'm sorry, I'm wrong it probably _should_ be `&&`.

Reply:
Hah, I know the problem, `Result` is missing an instance of `Resolve`.

Reply:
Ok, no. I reread the specification and understand the problem and also why I initially told you to use `==>`. 

The specification universally quantifies over a borrow, but what you want to state would in english be:

"For all mutable borrows of T, if the current value of `self` is `*t` then the final value of `self` is `^t`, but this _cannot_ be true, there's no reason why this would be true of _any_ borrow. Instead, what you want is an existential quantification: 

`exists<t : &mut T> *self == Ok(*t) && ^self == Ok(^t) && result == Ok(t) || ...`

Reply:
That does work! So the custom implementation was worth it.

Reply:
Yep! I know its tedious, but here's value in doing the proofs :)